guardrails/utils/telemetry_utils.py

```diff
@@ -551,3 +429,650 @@ def trace_llm_call(
+    #     (
+    #         norm(raw_embed_np, axis=0)*norm(validated_embed_np, axis=0)
+    #     )
+    # )
+
+    # input_x_validated_output_cosine = (
+    #     np.sum(input_embed_np*validated_embed_np, axis=0)
+    #     /
+    #     (
+    #         norm(input_embed_np, axis=0)*norm(validated_embed_np, axis=0)
+    #     )
+    # )
+
+    # input_x_raw_output_cosine = (
+    #     np.sum(input_embed_np*raw_embed_np, axis=0)
+    #     /
+    #     (
+    #         norm(input_embed_np, axis=0)*norm(raw_embed_np, axis=0)
+    #     )
+    # )
+
+    # guard_span.set_attribute(
+    #     "raw_output_x_validated_output_cosine",
+    #     float(str(raw_output_x_validated_output_cosine))
+    # )
+    # guard_span.set_attribute(
+    #     "input_x_validated_output_cosine",
+    #     float(str(input_x_validated_output_cosine))
+    # )
+    # guard_span.set_attribute(
+    #     "input_x_raw_output_cosine",
+    #     float(str(input_x_raw_output_cosine))
+    # )
+
+
+def trace_stream_guard(
+    guard_span: Span,
+    result: Iterable[ValidationOutcome[OT]],
+    history: Stack[Call],
+) -> Iterable[ValidationOutcome[OT]]:
+    next_exists = True
+    while next_exists:
+        try:
+            res = next(result)  # type: ignore
+            # FIXME: This should only be called once;
+            # Accumulate the validated output and call at the end
+            add_guard_attributes(guard_span, history, res)
+            yield res
+        except StopIteration:
+            next_exists = False
+
+
+def trace_guard_execution(
+    guard_name: str,
+    history: Stack[Call],
+    _execute_fn: Callable[
+        ..., Union[ValidationOutcome[OT], Iterable[ValidationOutcome[OT]]]
+    ],
+    tracer: Optional[Tracer] = None,
+    *args,
+    **kwargs,
+) -> Union[ValidationOutcome[OT], Iterable[ValidationOutcome[OT]]]:
+    if not settings.disable_tracing:
+        current_otel_context = context.get_current()
+        tracer = tracer or trace.get_tracer("guardrails-ai", GUARDRAILS_VERSION)
+
+        with tracer.start_as_current_span(
+            name="guard",  # type: ignore
+            context=current_otel_context,  # type: ignore
+        ) as guard_span:
+            guard_span.set_attribute("guardrails.version", GUARDRAILS_VERSION)
+            guard_span.set_attribute("type", "guardrails/guard")
+            guard_span.set_attribute("guard.name", guard_name)
+
+            try:
+                result = _execute_fn(*args, **kwargs)
+                if isinstance(result, Iterable) and not isinstance(
+                    result, ValidationOutcome
+                ):
+                    return trace_stream_guard(guard_span, result, history)
+                add_guard_attributes(guard_span, history, result)
+                return result
+            except Exception as e:
+                guard_span.set_status(status=StatusCode.ERROR, description=str(e))
+                raise e
+    else:
+        return _execute_fn(*args, **kwargs)
```

Comment:
did not expect the non traced handling to be here in the trace_guard_execution but not the end of the world

Reply:
The reason we're checking the settings here vs where we're wrapping the call is really just verbosity.  If we check in `guard.py` we'd have to repeat all the args against whereas here they're captured as spreads (or w/e we call those in python).